File: Java/D12/d12p3.java
/*
 * A tenth standard student has been given a task, Given P number of subject marks 
and a number I, He has to print the I-th least value of sums among all the 
possible sub-arrays of marks.

Input Format:
-------------
Line-1: Two space separated inetegers, P and I.
Line-2: P space separated integers, marks[].

Output Format:
--------------
Print the I-th least value of possible sums.



Sample Input-1:
--------------
3 4
3 2 4

Sample output-1:
---------------
5

Explanation: 
------------
The subarrays of 3 2 4 are:
1st subarray: 3 the sum is 3
2nd subarray: 2 the sum is 2
3rd subarray: 4 the sum is 4
4th subarray: 3,2 the sum is 5
5th subarray: 2,4 the sum is 6
6th subarray: 3,2,4 the sum is 9

The 4th smallest is 5


Sample Input-2:
---------------
4 7
2 2 4 4

Sample output-2:
----------------
8


Explanation: 
------------
The subarrays of 2 2 4 4 are

1st subarray: 2 the sum is 2
2nd subarray: 2 the sum is 2
3rd subarray: 4 the sum is 4
4th subarray: 4 the sum is 4
5th subarray: 2,2 the sum is 4
6th subarray: 2,4 the sum is 6
7th subarray: 4,4 the sum is 8
8th subarray: 2,2,4 the sum is 8
9th subarray: 2,4,4 the sum is 10
10th subarray: 2,2,4,4 the sum is 8

The 7th smallest is 8
 */

import java.util.*;

/**
 * d12p3
 */
public class d12p3 {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int p = sc.nextInt();
        int i = sc.nextInt();
        int[] marks = new int[p];
        for (int j = 0; j < p; j++) {
            marks[j] = sc.nextInt();
        }
        sc.close();
        int[] sums = new int[p * (p + 1) / 2];
        int k = 0;
        for (int j = 0; j < p; j++) {
            int sum = 0;
            for (int l = j; l < p; l++) {
                sum += marks[l];
                sums[k++] = sum;
            }
        }
        Arrays.sort(sums);
        System.out.println(sums[i - 1]);
    }
    
}

File: CPP/D12/d12p3.cpp
// There are n boxes of balls in a row, every i-th box contains balls[i] balls.

// A series of boxes is called Arithmetic, if it consists of at least three boxes
// and if the difference between the number of balls in any two consecutive boxes 
// is same.

// A subset of boxes is called Arithmetic Subset,
// if subset[i], subset[i+1], subset[i+2], ..., subset[j] is Arithmetic.

// Your task is to findout the number of Arithmetic Subsets in the row of boxes.

// Input Format:
// -------------
// Line-1 : An integer N, number of boxes.
// Line-2 : N space separated integers balls[i], i-th box contains balls[i] balls. 

// Output Format:
// --------------
// Print an integer, number of Arithmetic Subsets.


// Sample Input:
// ---------------
// 4
// 1 2 3 4

// Sample Output:
// ----------------
// 3

// Explanation:
// ------------
// Arithmetic Subsets are: [1,2,3], [2,3,4], [1,2,3,4].


// Sample Input:
// ---------------
// 5
// 1 3 5 7 9

// Sample Output:
// ----------------
// 6

// Explanation:
// ------------
// Arithmetic Subsets are: [1,3,5], [3,5,7], [5,7,9], [1,3,5,7], [3,5,7,9], [1,3,5,7,9].


#include<bits/stdc++.h>
using namespace std;

int main(){
    int n;
    cin>>n;
    vector<int> nums(n,0);
    for(int i=0;i<n;++i){
        cin>>nums[i];
    }
    vector<int> dp(n,0);
    for(int i=2;i<n;++i){
        if(nums[i]-nums[i-1]==nums[i-1]-nums[i-2]){
            dp[i]+=dp[i-1]+1;
        }
    }
    int ans = 0;
    for(int i=0;i<n;++i){
        ans += dp[i];
    }
    cout<<ans;
}